### PR TITLE
Add testing scenario infrastructure

### PR DESCRIPTION
**Pull request by dbolharovcmx**

### CODE REVIEW

**Comment by dbolharovcmx** [comment]
[Create initial infrastructure for agent scenarios testing](https://trello.com/c/kqwFKPsC/1076-create-initial-infrastructure-for-agent-scenarios-testing)

File: tests/customport/bundle-overrides/volumes/traefik/http/traefik.toml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+defaultEntryPoints = ["http"]
+
+[entryPoints]
+[entryPoints.http]
+address = ":5001"
```
Comment (SnirCm) [inline tests/customport/bundle-overrides/volumes/traefik/http/traefik.toml:5]: this should stay 80

File: tests/customport/bundle-overrides/.env.defaults [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ACCESS_CONTROL_HOST=platform.checkmarx.net
+SCA_API_PROTOCOL=https
+SCA_API_HOST=api-sca.checkmarx.net
+
+# For https set to https
+TRANSFER_PROTOCOL=http
+EXTERNAL_HOSTNAME=127.0.0.1
+
+# For https set to 443
+TRAEFIK_PORT=5001
```
Comment (SnirCm) [inline tests/customport/bundle-overrides/.env.defaults:14]: this too should stay 80 it is the port inside the container

File: tests/customport/bundle-overrides/.env.defaults [hunk 3]
```diff
@@ -0,0 +1,48 @@
+
+########################### Traefik ###############################
+EXTERNAL_PORT=5001
```
Comment (SnirCm) [inline tests/customport/bundle-overrides/.env.defaults:3]: changing this to 5001 is all that is needed

File: tests/README.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
+# Scenario Testing
```
Comment (SnirCm) [inline tests/README.md:1]: 👌

**Review by SnirCm — CHANGES_REQUESTED** [review]

**Comment by dbolharovcmx** [comment]
[Add basic connectivity tests to the agent](https://trello.com/c/NXeVaCwL/1077-add-basic-connectivity-tests-to-the-agent)

**Comment by dbolharovcmx** [comment]
[Add HTTPS test to the agent](https://trello.com/c/UCpwJGXq/1078-add-https-test-to-the-agent)

File: tests/https/bundle-overrides/volumes/traefik/ssl/cert.crt [hunk 5]
```diff
@@ -0,0 +1,33 @@
+-----BEGIN CERTIFICATE-----
```
Comment (Snirx1) [inline tests/https/bundle-overrides/volumes/traefik/ssl/cert.crt:1]: self signed?

Comment (dbolharovcmx) [reply tests/https/bundle-overrides/volumes/traefik/ssl/cert.crt:1]: Yes, generated myself

**Review by Snirx1 — APPROVED** [review]

**Review by yellowblood — CHANGES_REQUESTED** [review]
.

**Review by maor-rosenfeld — CHANGES_REQUESTED** [review]
1. Move the python tests source code (and its Dockerfile) to the "tests" folder, because these tests are shared between the scenarios.
2. Avoid duplicating the entire env vars, instead only specify the vars which you're actually overriding

File: Jenkinsfile [hunk 6]
```diff
@@ -52,6 +52,70 @@ pipeline {
...
+
+                        files.each {
+
+                           def (testName, composeFile) = it.path.split('/')
+                           def testComposeFilePath = "tests/${testName}/${composeFile}"
+                           stash includes: "**", name: "tests"
+
+                           testingScenarios["test-${testName}"] = {
+                                node("docker"){
+                                    ws("${testName}-workspace"){
```
Comment (maor-rosenfeld) [inline Jenkinsfile:76]: 👌

File: Jenkinsfile [hunk 7]
```diff
@@ -52,6 +52,70 @@ pipeline {
...
+                                            {
+                                                sh label: "Apply Environment Variables", script: "echo -e \"\\n\" >> .env; for line in \$(cat .env.overrides); do echo \"\$line\" >> .env; done"
+                                            }
+
+                                            sh label: "Run Agent", script: "docker-compose -f docker-compose.yml up -d"
+                                            sh label: "Run Test", script: "docker-compose -f ${testComposeFilePath} up --build --abort-on-container-exit"
+
+                                            sh label: "Shutdown Test", script: "docker-compose -f ${testComposeFilePath} down --remove-orphans"
+                                            sh label: "Shutdown Agent", script: "docker-compose -f docker-compose.yml down"
+                                        }
```
Comment (maor-rosenfeld) [inline Jenkinsfile:108]: Can you move all this to a script file under the tests folder? Preferably a shell script file that you can also run locally without Jenkins. Easier to fix problems when you can run the same thing locally.

File: tests/README.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+variables, and some configuration for volumes you should process like this:
+```
+|--tests/
+   |--{test directory}/
+      |--bundle-overrides/
+         |--volumes/
+            |--traefik/
+               |--http/
+                  |--...
+         |--.env.defaults
```
Comment (maor-rosenfeld) [inline tests/README.md:42]:
```suggestion
         |--.env
```

File: tests/README.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+         |--docker-compose.yml
+         |--...
+      |--...
+```
+
+Files and directories inside the `bundle-overrides` will simply replace original SCA Agent's files during the runtime
+
+#### Override Environment Variables
+Environment Variables can be overwritten by adding `.env.overrides` files in the root of `bundle-overrides`.
```
Comment (maor-rosenfeld) [inline tests/README.md:51]: why not just ".env"?
```suggestion
Environment Variables can be overwritten by adding `.env` files in the root of `bundle-overrides`. 
```

File: tests/customport/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+version: '3.7'
+
+services:
+  customport-test:
+    image: customport-check-agent
+    build: ../src/connectivity
+    network_mode: host
+    environment:
+      AGENT-PORT: 5001
```
Comment (maor-rosenfeld) [inline tests/customport/docker-compose.yml:9]: A dash in env vars is a big no no
```suggestion
      AGENT_PORT: 5001
```

**Review by maor-rosenfeld — COMMENTED** [review]
How do you run specific tests for specific scenarios?

File: tests/README.md (hunk 9 above)
Comment (dbolharovcmx) [reply tests/README.md:51]: The thing is that then `Setup.sh` is adding `.env` file during the install and there is no automatic merge for already existing `.env` file. So the `.env.overrides` just adds then the contents to the end overwritting the previous value. Not elegant at all, but no need to implement the logic on finding and replacing the values.

**Comment by dbolharovcmx** [comment]
[Agent Test Infra - Refactoring](https://trello.com/c/EUyZcwNo/1104-agent-test-infra-refactoring)

**Review by maor-rosenfeld — APPROVED** [review]

**Pull request merged**